developer/appdev/articles/03-service-builder/03-creating-o-r-map/99-migrating-a-service-builder-module-from-spring-di-to-osgi-ds.markdown

```diff
@@ -107,26 +107,44 @@ Next, you'll work out any remaining references you need.
 
 ## Step 3: Resolve Any Circular Dependencies
 
-Service Builder avoids adding references that could cause circular
-dependencies.
+A circular dependency between two modules occurs when each module (directly or
```

Comment:
We really need to be clear. Modules do not depend on each other in the same way services depend on each other.  We're talking about service dependencies only.

Reply:
Modules have dependencies. Module dependencies are usually related to package versions, modules fail during startup if there's no other module that has the package with the required version that is required.

Services do not fail the module deployment.  Services wait until all their references are satisfied before becoming active.  We have system:check which does various helpful checks. One of those checks sees if anything is not able to become active because of a missing required reference.

Reply:
With spring + Liferay, all spring beans in the same module act as a single bundle of services that become active together and are allowed to bind together before activation.  This is a Spring feature that DS does not have.  With DS all components are their own service and must be able to resolve on their own.